### PR TITLE
[GCS] Adding new integration for Custom GCS Input

### LINKED ISSUE #4330: Custom GCS input package
This issue is to track the creation of a custom input package for the new GCS input released in 8.5

### PR DESCRIPTION
**Pull request by P1llus**
## What does this PR do?

This adds an integration for the new Google Cloud Storage input that was added to filebeat in 8.5.0

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [ ]

## Related issues

- Closes #4330

### CODE REVIEW

**Comment by P1llus** [comment]
There is currently a few questions up for discussion here:

1. Will we add system tests, and how we should do it.
2. Which ECS fields should we use, continuing a discussion around if we should include at least fields related to common metadata processors.

**Comment by P1llus** [comment]
Integration has not been fully tested yet, but will be soon (before merging)

**Comment by P1llus** [comment]
Waiting for this to merge to add system tests: https://github.com/elastic/stream/pull/46

File: packages/google_cloud_storage/data_stream/generic/manifest.yml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        description: "This attribute contains the details about a specific bucket like name, max_workers, poll, poll_interval and bucket_timeout. The attribute name is specific to a bucket as it describes the bucket name, while the fields max_workers, poll, poll_interval and bucket_timeout can exist both at the bucket level and the root level. \nIt is internally represented as an array, so multiple buckets can be provided.\nFor more information about each attribute, please see the relevant [Documentation](https://www.elastic.co/guide/en/beats/filebeat/8.5/filebeat-input-gcs.html#attrib-buckets).\n"
+        required: true
+        show_user: true
+        default: |
+          - name: gcs-bucket1
+            max_workers: 3
+            poll: true
+            poll_interval: 15s
+            bucket_timeout: 60s
+            parse_json: false
```
Comment (kcreddy) [inline packages/google_cloud_storage/data_stream/generic/manifest.yml:59]: `parse_json` seems to be a root level field as per the [doc here](https://www.elastic.co/guide/en/beats/filebeat/master//filebeat-input-gcs.html#:~:text=file_path%7D%7D/%7B%7Bcreds_file_name%7D%7D.json-,parse_json%3A%20true,-buckets%3A%0A%20%20%2D)

File: packages/google_cloud_storage/data_stream/generic/agent/stream/gcs.yml.hbs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+{{#if project_id}}
+project_id: {{project_id}}
+{{/if}}
+{{#if service_account_key}}
+auth.credentials_json.account_key: {{service_account_key}}
+{{/if}}
+{{#if service_account_file}}
+auth.credentials_file.path: {{service_account_file}}
+{{/if}}
+{{#if buckets}}
```
Comment (kcreddy) [inline packages/google_cloud_storage/data_stream/generic/agent/stream/gcs.yml.hbs:15]: Another field `parse_json` could be added [as per the doc](https://www.elastic.co/guide/en/beats/filebeat/master//filebeat-input-gcs.html#:~:text=file_path%7D%7D/%7B%7Bcreds_file_name%7D%7D.json-,parse_json%3A%20true,-buckets%3A%0A%20%20%2D)

**Review by kcreddy — APPROVED** [review]
Just minor change requested. Rest LGTM 👍🏼

**Comment by narph** [comment]
@P1llus , can we get this in?

**Comment by ShourieG** [comment]
@narph @P1llus I feel if we could wait till this PR: https://github.com/elastic/beats/pull/34155 is merged, it would be awesome. This contains a lot of significant improvements to the GCS input which makes it more stable and behave better at scale. Having the integration live after the changes have gone through would be ideal.

**Comment by P1llus** [comment]
This was waiting for https://github.com/elastic/stream/pull/46 and https://github.com/elastic/elastic-package/pull/1073
The latter was only merged a few days ago, so I will go ahead and finish the first one.

The reason behind it was so that we could implement the dynamic ECS template already from the initial version. It also seems that we should bump the minimum version as per @ShourieG comment, so we can ensure we have all the features needed for the input to work as intended.

**Comment by P1llus** [comment]
All blocking features have now been resolved, will just have to update the PR with the system test and add dynamic ECS template before merging: https://github.com/elastic/integrations/issues/5055

**Comment by ShourieG** [comment]
@narph @P1llus PR: https://github.com/elastic/beats/pull/34155 is now merged, hence unblocking this PR.

**Comment by P1llus** [comment]
This new input needs to be added to the filebeat spec file for elastic-agent first: https://github.com/elastic/elastic-agent/blob/main/specs/filebeat.spec.yml

**Comment by P1llus** [comment]
Spec file updated as per: https://github.com/elastic/elastic-agent/pull/2149
Awaiting new snapshot image to be generated.

**Comment by P1llus** [comment]
There is currently no possible way to override the URL endpoint for the GCS client, we would need to add a `alternative_host` variable, similar to gcp-pubsub, so that we have the possibility to route the client to our local system test.

File: packages/google_cloud_storage/_dev/build/docs/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+1.  When processing google cloud buckets, if suddenly there is any outage, the process will be able to resume post the last file it processed and was successfully able to save the state for. 
+
+2.  If any errors occur for certain files, they will be logged appropriately, but the rest of the 
+    files will continue to be processed normally. 
+
+3.  If any major error occurs which stops the main thread, the logs will be appropriately generated,
+    describing said error.
+
+
+NOTE: Currently only `JSON` is supported with respect to object/file formats. As for authentication types, we currently have support for
```
Comment (ShourieG) [inline packages/google_cloud_storage/_dev/build/docs/README.md:19]:
```suggestion
NOTE: Currently only `JSON` is supported with respect to object/file formats. We also support gzipped JSON object/files. As for authentication types, we currently have support for 
```

File: packages/google_cloud_storage/data_stream/generic/agent/stream/gcs.yml.hbs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+tags:
+{{#each tags as |tag i|}}
+  - {{tag}}
+{{/each}}
+{{/if}}
+{{#contains "forwarded" tags}}
+publisher_pipeline.disable_host: true
+{{/contains}}
+{{#if processors}}
+processors:
```
Comment (ShourieG) [inline packages/google_cloud_storage/data_stream/generic/agent/stream/gcs.yml.hbs:32]:
```suggestion
processors:
{{processors}}
```

**Review by ShourieG — APPROVED** [review]
LGTM

**Pull request merged**